src/Titanium.Web.Proxy/Helpers/HttpHelper.cs

```diff
@@ -143,37 +145,23 @@ internal static Task<int> IsPriMethod(ICustomStreamReader clientStreamReader)
         /// <returns>
         ///     1: when starts with the given string, 0: when valid HTTP method, -1: otherwise
         /// </returns>
-        private static async Task<int> startsWith(ICustomStreamReader clientStreamReader, string expectedStart)
+        private static async Task<int> startsWith(ICustomStreamReader clientStreamReader, string expectedStart, CancellationToken cancellationToken = default(CancellationToken))
         {
-            bool isExpected = true;
+            int iRet = -1;
             int lengthToCheck = 10;
-            for (int i = 0; i < lengthToCheck; i++)
-            {
-                int b = await clientStreamReader.PeekByteAsync(i);
-                if (b == -1)
-                {
-                    return -1;
-                }
-
-                if (b == ' ' && i > 2)
-                {
-                    return isExpected ? 1 : 0;
-                }
 
-                char ch = (char)b;
-                if (!char.IsLetter(ch))
-                {
-                    return -1;
-                }
+            var vBuffer = await clientStreamReader.PeekBytesAsync(0, lengthToCheck, cancellationToken);
```

Comment:
I not really like this refactoring. Why is it needed?
Is any bug fixed in this change?

My problem is that it causes multipe new object allocations:
PeekBytesAsync : new byte[]
GetString: new string
Regex.. at least it shoudl be compiled, but why, when we just want to check only english A-Z characters?

Reply:
IS the multipel PeekBytesAsync slow? In that case maybe there should be an overload for PeekBytesAsync, which accepts a buffer, and fills that. Like the Read methods.

Reply:
I think await state-machine on loop is inefficient.
I'm of course ok for passing a buffer (we could use IBufferPool then).
The initial rule is (3 A-Z + space) but as I said, we never use the fact the method is HTTP, we could only write this :


    int iRet = -1;
    const int lengthToCheck = 10;
    byte[] buffer = null;
    try
    {
        buffer = bufferPool.GetBuffer(lengthToCheck);
        int iPeeked = await clientStreamReader.PeekBytesAsync(buffer, 0, lengthToCheck, cancellationToken);
        if (iPeeked > 0)
        {
            var httpMethod = defaultEncoding.GetString(buffer, 0, iPeeked);
            iRet = httpMethod.StartsWith(expectedStart) ? 1 : -1;
        }
    }
    finally
    {
        bufferPool.ReturnBuffer(buffer);
    }
    return iRet;
    //and why not give up the int ret -1/1 to bool

Reply:
I rewrote to have the best of the 2 worlds : use BufferPool with no string alloc and of course without looping on the await PeekByte. Let me know if it's ok for you.